app/server/lib/GristJobs.ts

```diff
@@ -0,0 +1,318 @@
+import { makeId } from 'app/server/lib/idUtils';
+import log from 'app/server/lib/log';
+import { Queue, Worker } from 'bullmq';
+import IORedis from 'ioredis';
+
+/**
+ *
+ * Support for queues.
+ *
+ * We use BullMQ for queuing, since it seems currently the best the
+ * node ecosystem has to offer. BullMQ relies on Redis. Since queuing
+ * is so handy, but we'd like most of Grist to be usable without Redis,
+ * we make some effort to support queuing without BullMQ. This
+ * may not be sustainable, we'll see.
+ *
+ * Important: if you put a job in a queue, it can outlast your process.
+ * That has implications for testing and deployment, so be careful.
+ *
+ */
+export interface IGristJobs {
+  // All workers and jobs are scoped to individual named queues,
+  // with the real interfaces operating at that level.
+  queue(queueName?: string): IGristQueueScope;
+}
+
+/**
+ * For a given queue, we can add jobs, or methods to process jobs,
+ */
+export interface IGristQueueScope {
+  /**
+   * Add a job.
+   */
+  add(name: string, data: any, options?: JobAddOptions): Promise<void>;
+
+
+  /**
+   * Add a job handler for all jobs regardless of name.
+   * Handlers given by handleName take priority, but no
+   * job handling will happen until handleDefault has been
+   * called.
+   */
+  handleDefault(defaultCallback: JobHandler): void;
+
+  /**
+   * Add a job handler for jobs with a specific name.
+   * Handler will only be effective once handleAll is called
+   * to specify what happens to jobs not matching expected
+   * names.
+   */
+  handleName(name: string,
+             callback: (job: GristJob) => Promise<any>): void;
+}
+
+/**
+ * The type of a function for handling jobs on a queue.
+ */
+export type JobHandler = (job: GristJob) => Promise<any>;
+
+/**
+ * The name used for a queue if no specific name is given.
+ */
+export const DEFAULT_QUEUE_NAME = 'default';
+
+/**
+ * BullMQ jobs are a string name, and then a data object.
+ */
+interface GristJob {
+  name: string;
+  data: any;
+}
+
+/**
+ * Options when adding a job. BullMQ has many more.
+ */
+interface JobAddOptions {
+  delay?: number;
+  jobId?: string;
+  repeat?: {
+    every: number;
+  }
+}
+
+/**
+ * Implementation for job functionality across the application.
+ */
+export class GristJobs implements IGristJobs {
```

Comment:
Just a suggestion. Jordi some time ago asked why we add those interfaces. It made me think, and I think he is right, it doesn't make much sense for the reader perspective.

What do you think about following convention:
- IGristJobs will be renamed to GristJobs (without the I)
- GristJobs will be renamed to GristRedisJobs or GristBullJobs

This way we have a protocol (or interface) for jobs, and a class that implements it in a specific way (in this instance over BullMq library). Adding another implementation, lets say in memmory will be easy, we just name it GristMemoryJobs, and so on.

It will also answer Jordi's question: why we need those interfaces. As it is easy to figure it out (there is a protocol and couple of implementations).

Reply:
So just renaming? I'm fine with that, was just following the rough existing convention.

Reply:
Yes, for now just renaming would be enough. Later on someone can do the split, and extract the memory part to a separate class.